File: apps/ios/App/ViewModels/NotificationsViewModel.swift
import Foundation
import Observation
import Factory
import Data
import Domain

/// The notification inbox — ported from apps/web/app/notifications/page.tsx.
///
/// The repository and the bell badge already existed; this is the screen the
/// badge was pointing at, which was a placeholder on both platforms.
///
/// Mirrors apps/android/.../ui/notifications/NotificationsViewModel.kt.
@Observable
@MainActor
final class NotificationsViewModel {
    @ObservationIgnored
    @Injected(\.notificationsRepository) private var notificationsRepository
    @ObservationIgnored
    @Injected(\.authRepository) private var authRepository

    private(set) var items: [NotificationRow] = []
    private(set) var unread = 0

    private var tasks: [Task<Void, Never>] = []
    private var userId: String?

    func start() {
        guard tasks.isEmpty else { return }
        tasks.append(Task { [weak self] in
            guard let self, let userId = await self.resolveUserId() else { return }
            self.userId = userId
            await withTaskGroup(of: Void.self) { group in
                group.addTask { await self.watchInbox(userId) }
                group.addTask { await self.watchUnread(userId) }
            }
        })
    }

    private func watchInbox(_ userId: String) async {
        do {
            for try await rows in try await notificationsRepository.watchInbox(userId: userId) {
                items = rows
            }
        } catch { print("Error watching notifications: \(error)") }
    }

    private func watchUnread(_ userId: String) async {
        do {
            for try await counts in try await notificationsRepository.watchUnreadCount(userId: userId) {
                unread = Int(counts.first ?? 0)
            }
        } catch { print("Error watching unread count: \(error)") }
    }

    func cancel() {
        tasks.forEach { $0.cancel() }
        tasks.removeAll()
    }

    func markRead(_ id: String) {
        Task { [weak self] in
            guard let self else { return }
            try? await self.notificationsRepository.markRead(id: id, at: nowIso())
        }
    }

    func markAllRead() {
        Task { [weak self] in
            guard let self, let userId = self.userId else { return }
            try? await self.notificationsRepository.markAllRead(userId: userId, at: nowIso())
        }
    }

    func dismiss(_ id: String) {
        Task { [weak self] in
            try? await self?.notificationsRepository.dismiss(id: id)
        }
    }

    /// Spelled out rather than `currentUserId ?? (try? await ensureUser())`:
    /// `??`'s right-hand side is `@autoclosure` and cannot hold an `await`.
    private func resolveUserId() async -> String? {
        if let existing = authRepository.currentUserId { return existing }
        return try? await authRepository.ensureUser()
    }
}

File: apps/android/app/src/main/java/com/sanvya/app/ui/notifications/NotificationsViewModel.kt
package com.sanvya.app.ui.notifications

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sanvya.app.data.auth.AuthRepository
import com.sanvya.app.data.repository.NotificationRow
import com.sanvya.app.data.repository.NotificationsRepository
import com.sanvya.app.data.repository.nowIso
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

/**
 * The notification inbox -- ported from apps/web/app/notifications/page.tsx.
 *
 * The repository and the bell badge already existed on both platforms; this is
 * the screen the badge was pointing at, which was a placeholder.
 *
 * Mirrors apps/ios/App/ViewModels/NotificationsViewModel.swift.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class NotificationsViewModel : ViewModel(), KoinComponent {
    private val notificationsRepository: NotificationsRepository by inject()
    private val authRepository: AuthRepository by inject()

    // flatMapLatest on the user id rather than a one-shot read: the queries are
    // scoped by user_id, and a guest who signs in mid-session must not keep
    // watching the id they had before.
    private val userId: StateFlow<String?> = authRepository.currentUserId

    val items: StateFlow<List<NotificationRow>> = userId.filterNotNull()
        .flatMapLatest { id -> notificationsRepository.watchInbox(id) }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val unread: StateFlow<Int> = userId.filterNotNull()
        .flatMapLatest { id -> notificationsRepository.watchUnreadCount(id) }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0)

    fun markRead(id: String) {
        viewModelScope.launch {
            val user = userId.filterNotNull().first()
            runCatching { notificationsRepository.markRead(user, id, nowIso()) }
        }
    }

    fun markAllRead() {
        viewModelScope.launch {
            val user = userId.filterNotNull().first()
            runCatching { notificationsRepository.markAllRead(user, nowIso()) }
        }
    }

    fun dismiss(id: String) {
        viewModelScope.launch { runCatching { notificationsRepository.dismiss(id) } }
    }
}
